### PR TITLE
Add `get_cuda_native_handle`

### LINKED ISSUE #564: [FEA]: Expose `get_cuda_native_handle` to Python
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cuda-python/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

cuda.bindings

### Is your feature request related to a problem? Please describe.

In #463 we added experimental C++ APIs `get_cuda_native_handle` for C/C++/Cython to access the C handles wrapped by cuda.bindings types/objects. This issue is to track the need of exposing such APIs to Python to replace the current usage of `__int__()`, something along this line:
```python
from cuda.bindings.utils import get_cuda_native_handle
...

err, s = driver.cuStreamCreate()
s_ptr = get_cuda_native_handle(s)
assert isinstance(s_ptr, int)
```
We may also consider moving the C++ counterparts from `cuda.core` to `cuda.bindings`.

### Describe the solution you'd like

n/a

### Describe alternatives you've considered

n/a

### Additional context

n/a

### PR DESCRIPTION
**Pull request by leofang**
## Description

closes #564

The performance is comparable to the existing `int()` approach (O(10) ns):
```python
In [1]: from cuda.bindings import driver

In [3]: s = driver.CUstream(1234)

In [4]: int(s)
Out[4]: 1234

In [6]: from cuda.bindings.utils import get_cuda_native_handle

In [7]: get_cuda_native_handle(s)
Out[7]: 1234

In [8]: %timeit int(s)
44.7 ns ± 1.34 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [9]: %timeit get_cuda_native_handle(s)
95.1 ns ± 1.06 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by leofang** [comment]
(added a short benchmark to the PR description)

**Comment by leofang** [comment]
/ok to test [8970dca](https://github.com/NVIDIA/cuda-python/pull/773/commits/8970dca6582fd2e8217c5d639158296dd1623245)

**Comment by leofang** [comment]
/ok to test [4ca551c](https://github.com/NVIDIA/cuda-python/pull/773/commits/4ca551c01600f204dbda65cc663944c73f9d52b1)

File: cuda_bindings/setup.py [hunk 1]
```diff
@@ -287,6 +288,7 @@ def prep_extensions(sources, libraries):
 
 # new path for the bindings from cybind
 def rename_architecture_specific_files():
+    path = os.path.join("cuda", "bindings", "_internal")
```
Comment (leofang) [inline cuda_bindings/setup.py:291]: Note: this is a bug fix exposed by the new line 224. Previously, the `path` variable happened to be right (leaking from the loop variable at line 227), but it's not longer correct after we append one more item to `path_list`.

**Comment by leofang** [comment]
Decision to make: Should we raise `DeprecationWarning` in `__int__()`? It can be very noisy but it's the only effective way for us to inform users about the deprecation.

File: cuda_bindings/tests/test_utils.py [hunk 2]
```diff
@@ -41,3 +44,41 @@ def test_ptx_utils(kernel, actual_ptx_ver, min_cuda_ver):
...
+    ),
+)
+def test_get_handle(target):
+    ptr = random.randint(1, 1024)
+    obj = target(ptr)
+    handle = get_cuda_native_handle(obj)
+    assert handle == ptr
+
+
+def test_get_handle_error():
```
Comment (rwgk) [inline cuda_bindings/tests/test_utils.py]: I'd `@pytest.mark.parametrize(...)` this too, although it doesn't reduce code size, to make the tests independent. And maybe add a couple extra corner cases (e.g. `None`, `()`, `{}`).

**Review by rwgk — APPROVED** [review]
> Decision to make: Should we raise `DeprecationWarning` in `__int__()`? It can be very noisy but it's the only effective way for us to inform users about the deprecation.

Do we want to give users a grace period, in case they need to support two cuda-bindings versions for a while?

Concretely:

In the next release, deprecate only in the documentation.

Only with the next minor release bump: actually raise `DeprecationWarning`.

File: cuda_bindings/tests/test_utils.py (hunk 2 above)
Comment (leofang) [reply cuda_bindings/tests/test_utils.py]: done in commit c467526

**Comment by rwgk** [comment]
> > Decision to make: Should we raise `DeprecationWarning` in `__int__()`? It can be very noisy but it's the only effective way for us to inform users about the deprecation.
> 
> Do we want to give users a grace period, in case they need to support two cuda-bindings versions for a while?
> 
> Concretely:
> 
> In the next release, deprecate only in the documentation.
> 
> Only with the next minor release bump: actually raise `DeprecationWarning`.

@leofang WDYT about introducing a "Phased deprecations" section somewhere, so we don't forget in the future.

E.g. for this case:

`__int__()` method for type A, B, C were informally deprecated in cuda-bindings release vX.Y.Z. `DeprecationWarning` will be raised starting with release vX.Y+1.0.

**Comment by leofang** [comment]
Yup that's great idea, tracked in #775.

**Comment by rwgk** [comment]
@leofang This PR is causing a weird issue under Windows. I first noticed when updating the `unreleased` branch, but the issue is actually showing up already in the CI logs here, e.g. this run against `main` right after this PR was merged:

* https://github.com/NVIDIA/cuda-python/actions/runs/16476401578/job/46579899006

All tests pass actually, but each invocation of `python` generates some noise. Have you seen something like this before?

```
Running bindinds tests
/c/a/cuda-python/cuda-python/cuda_bindings
Error processing line 4 of C:\a\_tool\Python\3.12.10\x64\Lib\site-packages\_cuda_bindings_redirector.pth:

  Traceback (most recent call last):
    File "<frozen site>", line 206, in addpackage
    File "<string>", line 1, in <module>
    File "C:\a\_tool\Python\3.12.10\x64\Lib\site-packages\_cuda_bindings_redirector.py", line 28, in <module>
      import cuda.bindings
    File "C:\a\_tool\Python\3.12.10\x64\Lib\site-packages\cuda\bindings\__init__.py", line 16, in <module>
      from cuda.bindings import utils
    File "C:\a\_tool\Python\3.12.10\x64\Lib\site-packages\cuda\bindings\utils\__init__.py", line 4, in <module>
      from ._get_handle import get_cuda_native_handle
    File "cuda\\bindings\\utils\\_get_handle.pyx", line 1, in init cuda.bindings.utils._get_handle
    File "cuda\\bindings\\_lib\\utils.pyx", line 1, in init cuda.bindings._lib.utils
    File "cuda\\bindings\\driver.pyx", line 1, in init cuda.bindings.driver
    File "cuda\\bindings\\cydriver.pyx", line 1, in init cuda.bindings.cydriver
    File "cuda\\bindings\\_bindings\\cydriver.pyx", line 6, in init cuda.bindings._bindings.cydriver
  ModuleNotFoundError: No module named 'win32api'

Remainder of file ignored
============================= test session starts =============================
platform win32 -- Python 3.12.10, pytest-8.4.1, pluggy-1.6.0 -- C:\a\_tool\Python\3.12.10\x64\python.exe
cachedir: .pytest_cache
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: C:\a\cuda-python\cuda-python\cuda_bindings\tests
configfile: pytest.ini
plugins: benchmark-5.1.0
collecting ... collected 176 items / 1 skipped

tests\test_cuda.py::test_cuda_memcpy PASSED                              [  0%]
... <snip> ...
================= 171 passed, 6 skipped, 3 warnings in 52.02s =================
```

**Comment by leofang** [comment]
I've seen errors locally like this too, but I thought it was because I had a corrupted environment and it did go away after a clean rebuild. I probably know what's going on, let me see if I can put together a patch quickly.

**Comment by rwgk** [comment]
> it did go away after a clean rebuild

That's surprising. I did multiple clean rebuilds (while backtracking from the `unreleased` testing), for me the issue was reliably reproducible.

> let me see if I can put together a patch quickly.

Sounds great, thanks!
